### PR TITLE
Consistent chart colors

### LINKED ISSUE #559: Stop chart colors from changing
We have received feedback from some users that the legend entries are confusing. An often repeated statement is that it is confusing that the colors in the charts and legends change when you select or deselect time series. 
As far as I know the color is determined by a "datasetIndex" (that changes, everytime the selection changes) and the corresponding position in the given list of colors. Based on this approach, I can't think of a way to fix a color to a time series, but maybe someone else has an idea about it?

### PR DESCRIPTION
**Pull request by brockfanning**
Fixes #559 

This should make the chart colors more consistent, but "remembering" the colors of disaggregation combinations. For example, if a user selects "Female" and it appears as a green line, and then later unselects "Female", and then later selects "Female" again, it should again show up as green.

Note that this "remembering" only lasts while the user is on the page. So, in the above example, if the user refreshes the page and selects "Female" again, there is no guarantee that it will appear as green again.

Disclaimer: This is one of the more (if not the most) complex chunks of code I've written for this project, so I would recommend thorough and critical testing.

Merging note - let's use "Squash" when merging this, as I accidentally included some notes...

### CODE REVIEW

**Comment by brockfanning** [comment]
@jwestw If you have time it might be nice to have a feature branch for this one. No special configuration is change needed besides pointing to the `brockfanning/consistent-chart-colors-take-2` branch.

**Comment by jwestw** [comment]
Feature branch: http://sdgdev-813006012.eu-west-1.elb.amazonaws.com/consistent-chart-colors-take-2

**Comment by jwestw** [comment]
I am looking at this indicator: 
http://sdgdev-813006012.eu-west-1.elb.amazonaws.com/consistent-chart-colors-take-2/5-3-2/

It seems to me that if you select and deselect a disagregation the same colour is automatically selected, so this is consistent. However I notice when deselecting, the colours change and sometimes another value takes on the colour of the deselected value. 

I am still playing around with that to see if I can show a clear image of what I am trying to explain here.

**Comment by brockfanning** [comment]
@jwestw I agree, it definitely seems to be broken on the feature branch. I must have missed something in the code, so I'll have to revisit.

**Comment by brockfanning** [comment]
@jwestw Actually I just pushed a simple change that should help, if you have time to refresh the feature branch and test again.

**Comment by jwestw** [comment]
I have rebuilt the feature branch but I am still seeing the problem. Again, I still looking at : http://sdgdev-813006012.eu-west-1.elb.amazonaws.com/consistent-chart-colors-take-2/5-3-2/

I am wondering if there's anything I have missed. Apart from updating the remote theme (https://github.com/ONSdigital/sdg-indicators/blob/4d53254af2fbfacadabe2f4effe83ce29721cac6/_config.yml#L149) is there anything else I should have updated?

**Comment by LucyGwilliamAdmin** [comment]
@jwestw are you sure your branch rebuilt? I can't see a new one in the slack channel

**Comment by jwestw** [comment]
This is a good point. I definitely re-ran the last action, so it would build. I struggled to get this to build yesterday. Not sure what the issue might be.

**Comment by jwestw** [comment]
Okay the feature branch rebuild and this is working now!

**Review by jwestw — APPROVED** [review]
Happy to approve on the basis of last comments.

**Comment by LucyGwilliamAdmin** [comment]
@jwestw I was just looking at your feature branch as wanted to see what new functionality was like but think there may be an issue:

1. Go to http://sdgdev-813006012.eu-west-1.elb.amazonaws.com/consistent-chart-colors-take-2/5-3-2/
2. Select all breakdowns under 'Age at which FGM took place'
3. Under 'Country of birth' select 'Born elsewhere in Africa', 'Born elsewhere is Asia' and 'Born in Eastern Africa'
4. Check colour for 'Born in Eastern Africa' in legend - teal green colour
5. Unselect 'Born in Eastern Africa'
6. Select 'Born in Northern Africa'
7. Check colour for 'Born in Northern Africa' in legend - also teal green colour

**Comment by jwestw** [comment]
I have tried to recreate this but the colours were consistent for me. I will continue to experiment.

**Comment by LucyGwilliamAdmin** [comment]
@jwestw for point 6, try other 'Country of birth' breakdowns

**Comment by brockfanning** [comment]
@LucyGwilliamAdmin I can't recreate with those steps -- for me in step 7 "Born in Northern Africa is red stripes - not teal green.

However what you describe sounds plausible based on this code, and I think is by design. When you get past the limit of disaggregations, and then unselect something, there has to be some way for the unselected color to be recycled (or "evicted" as I call it in the code). Otherwise there would be duplicate colors when you select something else after. This is the main reason that this feature was so confusing to conceptualize.

In fact I can recreate with these steps:

1. Go to http://sdgdev-813006012.eu-west-1.elb.amazonaws.com/consistent-chart-colors-take-2/5-3-2/
2. Select all breakdowns under 'Age at which FGM took place'
3. Under 'Country of birth' select 'Born elsewhere in Africa', 'Born elsewhere is Asia', 'Born in Eastern Africa', **and 'Born in Northern Africa'**
4. Check colour for 'Born in Eastern Africa' in legend - teal green colour
5. Unselect 'Born in Eastern Africa'
6. Select '**Born in UK**'
7. Check colour for '**Born in UK**' in legend - also teal green colour

But I think this is unavoidable because of the "color eviction" thing described above.

**Comment by LucyGwilliamAdmin** [comment]
@brockfanning ah ok

Also has this changed it so that when breakdowns are added, only the breakdown by itself is added and not the combination?

**Comment by brockfanning** [comment]
@LucyGwilliamAdmin Not intentionally - but let me know what you're seeing and I'll see if I can track down the problem in the code.

**Comment by LucyGwilliamAdmin** [comment]
@brockfanning It actually looks like it's specific to that indicator as it behaves the same on live site but if you:

1. Go to https://sdgdata.gov.uk/1-2-1/
2. Select 'Female' under 'Sex' - chart is updated with data for 'Female'
3. Select '15 and under' under 'Age' - chart is updated with data for '15 and under' *and* '15 and under, Female'

Whereas for 5.3.2
1. Go to https://sdgdata.gov.uk/5-3-2/
2. Select '1 to 4' under 'Age at which FGM took place' - chart is updated with '1 to 4'
3. Select 'Born elsewhere in Africa' under 'Country of birth' - chart is updated only with 'Born elsewhere in Africa' and not 'Born elsewhere in Africa, 1 to 4'

However, I only brought this up here as thought it was related as I remember discussing it at same time in a recent tech call, so could pick up elsewhere

**Comment by jwestw** [comment]
Just to reiterate and confirm: I am seeing consistent chart colours which is what this PR is aiming to fix. It's concerning that Lucy is able to get inconsistent colour though.

**Comment by LucyGwilliamAdmin** [comment]
@jwestw even when following Brock's recreation?

**Comment by brockfanning** [comment]
@LucyGwilliamAdmin That behavior means that the data for 5.3.2 doesn't have any rows that are both "1 to 4" and "Born elsewhere in Africa".

@jwestw Yes, it is probably an edge case, insofar as most users will probably not get to the maximum number of colors. However it's certainly something that will happen.

To reiterate though, this is unavoidable, due to the simple fact that we are working with a limited number of colors (eg, 12) and an unlimited number of combinations of disaggregations. In other words, it's impossible to assign one color to each combination. The best we can do is assign combinations to colors as the user picks them, use those assignments as long as possible, and then when we're out of colors, to re-assign colors are they're made available by unselected disaggregations.

To sum up the use-cases affected by this PR:

1. Users looking at indicators with less than 12 disaggregation combinations:  The color assignments will be 100% consistent.
2. Users looking at indicators with more than 12 disaggregation combinations, but who never select more than 12: The color assignments will be 100% consistent.
3. Users looking at indicators with more than 12 disaggregation combinations, and who select more than 12: Some colors will be re-used as the user deselects and selects combinations. At any given time, all 12 colors will appear with no duplication.
